modules/assim.sequential/R/sda.enkf_parallel.R

```diff
@@ -0,0 +1,644 @@
+                                        "      require (PEcAn.uncertainty)",
+                                        "      require (foreach)", 
+                                        "      qsub_sda_batch('@FOLDER_PATH@')", 
+                                        "    \" | R --no-save")
+                             jobsh <- gsub("@FOLDER_PATH@", folder.path, jobsh)
+                             writeLines(jobsh, con = file.path(folder.path, "job.sh"))
+                             # qsub command.
+                             qsub <- qsub.cmd
+                             if (grepl("NAME", qsub.cmd, fixed = TRUE)) {
+                               qsub <- gsub("@NAME@", paste0("Job-", i), qsub)
+                             }
+                             if (grepl("STDOUT", qsub.cmd, fixed = TRUE)) {
+                               qsub <- gsub("@STDOUT@", file.path(folder.path, "stdout.log"), qsub)
+                             }
+                             if (grepl("STDERR", qsub.cmd, fixed = TRUE)) {
+                               qsub <- gsub("@STDERR@", file.path(folder.path, "stderr.log"), qsub)
+                             }
+                             if (grepl("CORES", qsub.cmd, fixed = TRUE)) {
+                               qsub <- gsub("@CORES@", cores, qsub)
+                             }
+                             qsub <- strsplit(qsub, " (?=([^\"']*\"[^\"']*\")*[^\"']*$)", perl = TRUE)
+                             cmd <- qsub[[1]]
+                             out <- system2(cmd, file.path(folder.path, "job.sh"), stdout = TRUE, stderr = TRUE)
+                           }
+}
+
+##' This function can help to execute sda function.
+##' @title qsub_sda_batch
+##' @param folder.path character: path where the `configs.rds` file is stored.
+##' @author Dongchen Zhang.
+##' @export
+qsub_sda_batch <- function(folder.path) {
+  configs <- readRDS(file.path(folder.path, "configs.rds"))
+  setting <- PEcAn.settings::read.settings(configs$setting)
+  sda.enkf_local(setting, 
+                 configs$obs.mean, 
+                 configs$obs.cov, 
+                 configs$Q, 
+                 configs$pre_enkf_params,
+                 configs$ensemble.samples,
+                 configs$outdir,
+                 configs$control)
+}
+
+##' This function can help to assemble sda outputs (analysis and forecasts) from each job execution.
+##' @title sda_assemble
+##' @param batch.folder character: path where the SDA batch jobs stored.
+##' @param outdir character: path where we want to store the assembled analysis and forecasts.
+##' @author Dongchen Zhang.
+##' @export
+sda_assemble <- function (batch.folder, outdir) {
+  # find folder paths to each SDA job.
+  folders <- list.files(batch.folder, full.names = T)
+  fail.folders <- (folders %>% furrr::future_map(function(f){
+    if(!file.exists(file.path(f, "sda.all.forecast.analysis.Rdata"))) {
+      return(f)
+    }
+  }) %>% unlist)
```

Comment:
This looks like it works fine, but FTR since file.exists accepts vectors I think you could skip the `map` entirely:

```suggestion
fail.folders <- folders[!file.exists(file.path(folders, "sda.all.forecast.analysis.Rdata"))]
```

Reply:
Thanks for the suggestion!